cohttp-lwt-unix/src/net.ml

```diff
@@ -21,26 +21,58 @@ open Lwt.Infix
 
 module IO = Io
 
-type ctx = {
-  ctx: Conduit_lwt_unix.ctx;
-  resolver: Resolver_lwt.t;
-} [@@deriving sexp_of]
+type ctx = (Conduit.resolvers[@sexp.opaque]) [@@deriving sexp]
 
-let init ?(ctx=Conduit_lwt_unix.default_ctx)
-         ?(resolver=Resolver_lwt_unix.system) () =
-  { ctx; resolver }
+let authenticator ~host:_ _ = Ok None
 
-let default_ctx = {
-  resolver = Resolver_lwt_unix.system;
-  ctx = Conduit_lwt_unix.default_ctx;
-}
+let tls_config =
+  Tls.Config.client ~authenticator ()
+
+let init ?(tls_config = tls_config) () =
+  Conduit_lwt.empty
+  |> Conduit_lwt.add Conduit_lwt.TCP.protocol
+    (Conduit_lwt.TCP.resolve ~port:80)
+  |> Conduit_lwt.add ~priority:10 Conduit_lwt_tls.TCP.protocol
+    (Conduit_lwt_tls.TCP.resolve ~port:443 ~config:tls_config)
+
+let default_ctx = init ()
+
+let failwith fmt = Format.kasprintf (fun err -> Lwt.fail (Failure err)) fmt
+
+let uri_to_endpoint uri =
+  (match Uri.host uri with
+   | None -> failwith "Invalid uri: no host component in %a" Uri.pp uri
+   | Some h -> Lwt.return h) >>= fun v ->
+  let ( >>= ) x f = match x with Ok x -> f x | Error err -> Error err in
+  match Domain_name.(of_string v >>= host), Ipaddr.of_string v with
+  | Ok domain_name, _ -> Lwt.return (Conduit.Endpoint.domain domain_name)
+  | Error _, Ok v -> Lwt.return (Conduit.Endpoint.ip v)
+  | Error _, Error _ -> failwith "Invalid uri: %a" Uri.pp uri
 
 let connect_uri ~ctx uri =
-  Resolver_lwt.resolve_uri ~uri ctx.resolver
-  >>= fun endp ->
-  Conduit_lwt_unix.endp_to_client ~ctx:ctx.ctx endp
-  >>= fun client ->
-  Conduit_lwt_unix.connect ~ctx:ctx.ctx client
+  uri_to_endpoint uri >>= fun edn ->
+  (* XXX(dinosaure): this situation is not the right one, [connect_uri] should
+   * take [Conduit.Endpoint.t] and use the given [ctx] which is well-defined
+   * by the user. However, [Cohttp] proposes a /default/ [ctx].
+   *
+   * We can see an another use of [Conduit] with [Cohttp_async] which does not
+   * rely on the given [ctx] but it uses [Conduit_async.connect] according to
+   * the introspection of the given [uri]. We have 3 incompatible choices here:
+   * - expect an user's [ctx] with a [Conduit.Endpoint.t]
+   * - make a [ctx] from the given [uri]
+   * - *)
+  let ctx = match Uri.scheme uri, Uri.port uri with
+    | Some "https", Some port ->
+      Conduit_lwt.add ~priority:0 Conduit_lwt_tls.TCP.protocol (Conduit_lwt_tls.TCP.resolve ~port ~config:tls_config) ctx
```

Comment:
What if `ctx = init ~tls_config:some_other_tls_config ()`? Why here we resolve always using the default config? Shouldn't we pass the same one as the context instead?

Reply:
The best for us (and to have an understandable behavior of `conduit`) is to construct a _special_ `ctx`/`Conduit.resolvers` from the given `Uri.t` - like, an introspection of the scheme, the port, etc. to produce the right way to initiate a connection. The `ctx` still will be on the control of the user - and can prioritize an other initialization (depending on his context) - but we should not try to use what `init ()` gives to us or `default_ctx` - I mean, such `ctx` does not correspond in anyway to the given `Uri.t` and it's why the current status of `cohttp` has some strange behaviors (which don't reflect the given `Uri.t`)